### PR TITLE
Flexmatch bans POST API.

### PR DESCRIPTION
**Pull request by y4ure**
- POST API for banning players for a duration.
- Updated BanInfo with custom_unban_date to support banning for a duration.
- Updated BanInfo with readonly bool to avoid locking.

### CODE REVIEW

File: driftbase/api/matchmakers/flexmatch.py [hunk 1]
```diff
@@ -272,6 +276,33 @@ def put(self):
...
+        player_ids = fields.List(fields.Integer(), required=True, metadata=dict(description="The list of player ids to ban."))
+        days = fields.Integer(validate=validate.Range(min=0))
+        hours = fields.Integer(validate=validate.Range(min=0))
+        minutes = fields.Integer(validate=validate.Range(min=0))
+        seconds = fields.Integer(validate=validate.Range(min=0))
+
+    @requires_roles("service")
+    @bp.arguments(BanArgs)
+    @bp.response(http_client.CREATED)
+    def post(self, args):
```
Comment (y4ure) [inline driftbase/api/matchmakers/flexmatch.py:291]: Currently, this bans players for all matchmakers.

File: driftbase/flexmatch.py [hunk 2]
```diff
@@ -853,20 +880,32 @@ def __init__(self, user_message, debug_info):
...
-        self._lock = self._redis.conn.lock(self._key + "LOCK", timeout=30) if self._redis else None
-        self._ttl = ttl or self.config.get("expiry_seconds", 60 * 60)
+        self._lock = self._redis.conn.lock(self._key + "LOCK", timeout=30) if (self._redis and not readonly) else None
+        self._ttl = config.get("expiry_seconds", 60 * 60)
 
     def __enter__(self):
-        self._lock.acquire(blocking=True)
-        value = g.redis.conn.get(self._key)
+        if self._lock:
+            self._lock.acquire(blocking=True)
```
Comment (y4ure) [inline driftbase/flexmatch.py:907]: Lock skipped when readonly.

File: driftbase/flexmatch.py [hunk 3]
```diff
@@ -880,10 +919,13 @@ def _post_value_loaded(self):
...
                 self._value["last_ban_date"] = datetime.fromisoformat(last_ban_date)
+            custom_unban_date = self._value.get("custom_unban_date")
+            if isinstance(custom_unban_date, str):
+                self._value["custom_unban_date"] = datetime.fromisoformat(custom_unban_date)
 
     def __exit__(self, exc_type, exc_val, exc_tb):
-        if self._lock.owned():
-            with g.redis.conn.pipeline() as pipe:
+        if self._lock is not None and self._lock.owned():
+            with self._redis.conn.pipeline() as pipe:
```
Comment (judgeaxl) [inline driftbase/flexmatch.py:928]: Would you need exception handling here to make sure the lock is always released even if the delete/set throws?

File: driftbase/flexmatch.py [hunk 4]
```diff
@@ -894,23 +936,48 @@ def __exit__(self, exc_type, exc_val, exc_tb):
     def is_banned(self) -> bool:
```
Comment (judgeaxl) [inline driftbase/flexmatch.py:936]: How come is_banned doesn't take the list of matches into account?

**Review by judgeaxl — CHANGES_REQUESTED** [review]
Feels like banning on match and banning on time conflict, is_banned don't check the match, for example, just the ban time.

**Comment by y4ure** [comment]
> Feels like banning on match and banning on time conflict, is_banned don't check the match, for example, just the ban time.

When banning from match, the unban date is calculated from the number of total bans and the "ban_time_seconds" read from the config. `get_unban_date` returns the greater of the calculated unban date or custom unban date.

File: driftbase/flexmatch.py (hunk 4 above)
Comment (y4ure) [reply driftbase/flexmatch.py:943]: is_banned checks get_unban_date() where it checks num_bans.
I'll update it to check the list of matches instead.

Comment (y4ure) [reply driftbase/flexmatch.py:943]: updated

**Review by judgeaxl — APPROVED** [review]

**Pull request merged**